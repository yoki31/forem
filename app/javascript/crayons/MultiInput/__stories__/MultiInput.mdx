# Multi-Input

The MultiInput component can be used in situations where a user wants to capture/add multiple entries.
It allows the ability to add, edit, remove and capture and validate each entry before submitting the form.

## Adding an entry

Once a user types their entry, they may submit it by pressing Enter, Comma or Space.

## Editing and removing an entry

When a user adds an entry, it will be displayed as a button group. The first button triggers an edit, and the second button triggers the removal of the entry.

Both buttons appear in the tab order, and are semantically grouped together for assistive technology.

## Technical implementation notes

### Styling

The component comes with some default styling for selected options, but the layout of selections can be customised by passing in a Preact component `SelectionTemplate`.

If the template is passed, it will receive in props the full option object in props e.g. `{ name: 'option one', someOtherProperty: 'foo' }`.

### Accessibility

Selected items appear as a button group with further context provided in `aria-label`s for screen reader users (Edit/Remove). As selections are made from the list, or removed from the selected items, they are announced to screen reader users via a visually hidden `aria-live` region.

### Additional

In the future, we may want to compare the MultiSelectAutocomplete Component, the AutocompleteTriggerTextArea and the MultiInput Component to look for abstractions and possibly move them into Custom Hooks.
