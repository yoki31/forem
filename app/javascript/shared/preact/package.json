{
  "name": "preact-root-fragment",
  "version": "0.2.0",
  "main": "./preact-root-fragment.js",
  "module": "./preact-root-fragment.js",
  "type": "module"
}
