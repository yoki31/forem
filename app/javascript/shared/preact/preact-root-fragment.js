/**
 * A Preact 11+ implementation of the `replaceNode` parameter from Preact 10.
 *
 * This creates a "Persistent Fragment" (a fake DOM element) containing one or more
 * DOM nodes, which can then be passed as the `parent` argument to Preact's `render()` method.
 * Source: https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
 */
export function createRootFragment(parent, replaceNode) {
  if (replaceNode) {
    replaceNode = Array.isArray(replaceNode) ? replaceNode : [replaceNode];
  } else {
    replaceNode = [parent];
    parent = parent.parentNode;
  }

  const s = replaceNode[replaceNode.length - 1].nextSibling;

  const rootFragment = {
    nodeType: 1,
    parentNode: parent,
    firstChild: replaceNode[0],
    childNodes: replaceNode,
    insertBefore: (c, r) => {
      parent.insertBefore(c, r || s);
      return c;
    },
    appendChild: (c) => {
      parent.insertBefore(c, s);
      return c;
    },
    removeChild: (c) => {
      parent.removeChild(c);
      return c;
    },
  };

  parent.__k = rootFragment;
  return rootFragment;
}
