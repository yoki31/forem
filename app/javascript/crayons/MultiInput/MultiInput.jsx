import { h, Fragment } from 'preact';
import PropTypes from 'prop-types';
import { useRef, useState, useEffect } from 'preact/hooks';
import { DefaultSelectionTemplate } from '../../shared/components/defaultSelectionTemplate';

const KEYS = {
  ENTER: 'Enter',
  COMMA: ',',
  SPACE: ' ',
  DELETE: 'Backspace',
};

/**
 * Component allowing users to add multiple entries for a given input field that get displayed as destructive pills
 *
 * @param {Object} props
 * @param {string} props.labelText The text for the input's label
 * @param {boolean} props.showLabel Whether the label text should be visible or hidden (for assistive tech users only)
 * @param {string} props.placeholder Input placeholder text
 * @param {string} props.inputRegex Optional regular expression used to restrict the input
 * @param {string} props.validationRegex Optional regular expression used to validate the value of the input
 * @param {Function} props.SelectionTemplate Optional Preact component to render selected items
 */

export const MultiInput = ({
  placeholder,
  inputRegex,
  validationRegex,
  showLabel = true,
  labelText,
  SelectionTemplate = DefaultSelectionTemplate,
}) => {
  const inputRef = useRef(null);
  const inputSizerRef = useRef(null);
  const selectedItemsRef = useRef(null);

  const [items, setItems] = useState([]);
  const [editValue, setEditValue] = useState(null);
  const [inputPosition, setInputPosition] = useState(null);

  useEffect(() => {
    // editValue defaults to null when component is first rendered.
    // This ensures we do not autofocus the input before the user has started interacting with the component.
    if (editValue === null) {
      return;
    }

    const { current: input } = inputRef;
    if (input && inputPosition !== null) {
      // Entering 'edit' mode
      resizeInputToContentSize();
      input.value = editValue;
      const { length: cursorPosition } = editValue;
      input.focus();
      // This will set the cursor position at the end of the text.
      input.setSelectionRange(cursorPosition, cursorPosition);
    }
  }, [inputPosition, editValue]);

  const handleInputBlur = ({ target: { value } }) => {
    addItemToList(value);
    clearInput();
  };

  const handleInputChange = async ({ target: { value } }) => {
    // When the input appears inline in "edit" mode, we need to dynamically calculate the width to ensure it occupies the right space
    // (an input cannot resize based on its text content). We use a hidden <span> to track the size.
    inputSizerRef.current.innerText = value;

    if (inputPosition !== null) {
      resizeInputToContentSize();
    }
  };

  const handleKeyDown = (e) => {
    const { value: currentValue } = inputRef.current;

    switch (e.key) {
      case KEYS.SPACE:
      case KEYS.ENTER:
      case KEYS.COMMA:
        e.preventDefault();
        addItemToList(e.target.value);
        clearInput();
        break;
      case KEYS.DELETE:
        if (currentValue === '') {
          e.preventDefault();
          editPreviousSelectionIfExists();
        }
        break;
      default:
        if (inputRegex && !inputRegex.test(e.key)) {
          e.preventDefault();
        }
    }
  };

  const addItemToList = (value) => {
    if (value.trim().length > 0) {
      // If an item was edited, we want to keep it in the same position in the list
      const insertIndex = inputPosition !== null ? inputPosition : items.length;

      // if we do not pass in a validationRegex we can assume that anything is valid
      const valid = validationRegex ? checkValidity(value) : true;
      const newSelections = [
        ...items.slice(0, insertIndex),
        { value, valid },
        ...items.slice(insertIndex),
      ];

      // We update the hidden selected items list, so additions are announced to screen reader users
      const listItem = document.createElement('li');
      listItem.innerText = value;
      selectedItemsRef.current.appendChild(listItem);

      setItems([...newSelections]);
      exitEditState({});
    }
  };

  const checkValidity = (value) => {
    return validationRegex.test(value);
  };

  const clearInput = () => {
    inputRef.current.value = '';
  };

  const resizeInputToContentSize = () => {
    const { current: input } = inputRef;

    if (input) {
      input.style.width = `${inputSizerRef.current.clientWidth}px`;
    }
  };

  const deselectItem = (clickedItem) => {
    const newArr = items.filter((item) => item.value !== clickedItem);
    setItems(newArr);

    // We also update the hidden selected items list, so removals are announced to screen reader users
    selectedItemsRef.current.querySelectorAll('li').forEach((selectionNode) => {
      if (selectionNode.innerText === clickedItem) {
        selectionNode.remove();
      }
    });
  };

  // If there is a previous selection, then pop it into edit mode
  const editPreviousSelectionIfExists = () => {
    if (items.length > 0 && inputPosition !== 0) {
      const nextEditIndex =
        inputPosition !== null ? inputPosition - 1 : items.length - 1;

      const item = items[nextEditIndex];
      enterEditState(item.value, nextEditIndex);
    }
  };

  const enterEditState = (editItem, editItemIndex) => {
    inputSizerRef.current.innerText = editItem;
    deselectItem(editItem);
    setEditValue(editItem);
    setInputPosition(editItemIndex);
  };

  const exitEditState = ({ nextInputValue = '' }) => {
    // Reset 'edit mode' input resizing
    inputRef.current?.style?.removeProperty('width');

    inputSizerRef.current.innerText = nextInputValue;
    setEditValue(nextInputValue);
    setInputPosition(nextInputValue === '' ? null : inputPosition + 1);
    // Blurring away while clearing the input
    if (nextInputValue === '') {
      inputRef.current.value = '';
    }
  };

  const allSelectedItemElements = items.map((item, index) => {
    // When we are in "edit mode" we visually display the input between the other selections
    // If the item being edited appears before the item being rendered then we set its position to
    // the index + 1 which matches the order, however, any items that appear after the item that is
    // being edited will need to increment their position by one to make place for the item being edited.

    // at this point the position is already set
    const defaultPosition = index + 1;
    const appearsBeforeInput = inputPosition === null || index < inputPosition;
    const position = appearsBeforeInput ? defaultPosition : defaultPosition + 1;
    return (
      <li
        key={index}
        className="c-input--multi__selection-list-item w-max"
        style={{ order: position }}
      >
        <SelectionTemplate
          name={item.value}
          className={`c-input--multi__selected ${
            !item.valid ? 'c-input--multi__selected-invalid' : ''
          }`}
          enableValidation={true}
          valid={item.valid}
          onEdit={() => enterEditState(item.value, index)}
          onDeselect={() => deselectItem(item.value)}
        />
      </li>
    );
  });

  return (
    <Fragment>
      <span
        ref={inputSizerRef}
        aria-hidden="true"
        className="absolute pointer-events-none opacity-0 p-2"
      />
      <label
        id="multi-select-label"
        className={showLabel ? '' : 'screen-reader-only'}
      >
        {labelText}
      </label>

      {/* A visually hidden list provides confirmation messages to screen reader users as an item is selected or removed */}
      <div className="screen-reader-only">
        <p>Selected items:</p>
        <ul
          ref={selectedItemsRef}
          className="screen-reader-only list-none"
          aria-live="assertive"
          aria-atomic="false"
          aria-relevant="additions removals"
        />
      </div>

      <div class="c-input--multi relative">
        <div class="c-input--multi__wrapper-border crayons-textfield flex items-center cursor-text pb-9">
          <ul class="list-none flex flex-wrap w-100">
            {allSelectedItemElements}
            <li
              class="self-center"
              style={{
                order:
                  inputPosition === null ? items.length + 1 : inputPosition + 1,
              }}
            >
              <input
                autocomplete="off"
                class="c-input--multi__input"
                type="text"
                aria-labelledby="multi-select-label"
                onBlur={handleInputBlur}
                onKeyDown={handleKeyDown}
                placeholder={inputPosition === null ? placeholder : null}
                onChange={handleInputChange}
                ref={inputRef}
              />
            </li>
          </ul>
        </div>
      </div>
    </Fragment>
  );
};

MultiInput.propTypes = {
  labelText: PropTypes.string.isRequired,
  showLabel: PropTypes.bool,
  placeholder: PropTypes.string,
  inputRegex: PropTypes.string,
  validationRegex: PropTypes.string,
  SelectionTemplate: PropTypes.func,
};
