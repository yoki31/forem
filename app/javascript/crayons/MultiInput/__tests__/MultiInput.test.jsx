import { h } from 'preact';
import { render, waitFor } from '@testing-library/preact';
import { userEvent } from '@testing-library/user-event';

import '@testing-library/jest-dom';

import { MultiInput } from '../MultiInput';

describe('<MultiInput />', () => {
  const getProps = () => ({
    labelText: 'Example label',
    placeholder: 'Add an email address...',
    regex: /([a-zA-Z0-9@.])/,
    validationRegex: /^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$/,
  });

  const renderMultiInput = (props = getProps()) =>
    render(<MultiInput {...props} />);

  const setup = (getByPlaceholderText) => {
    const input = getByPlaceholderText('Add an email address...');
    input.focus();
    return input;
  };

  it('renders default UI', () => {
    const { container } = renderMultiInput();
    expect(container.innerHTML).toMatchSnapshot();
  });

  describe('different keys', () => {
    it('adds selection when enter is pressed', async () => {
      const { getByPlaceholderText, getByRole } = renderMultiInput();
      const input = setup(getByPlaceholderText);

      await waitFor(() => userEvent.type(input, 'forem@gmail.com'));
      userEvent.type(input, '{enter}');
      await waitFor(() =>
        expect(
          getByRole('button', { name: 'Edit forem@gmail.com' }),
        ).toBeInTheDocument(),
      );
    });

    it('adds selection when comma is pressed', async () => {
      const { getByPlaceholderText, getByRole } = renderMultiInput();
      const input = setup(getByPlaceholderText);

      await waitFor(() => userEvent.type(input, 'forem@gmail.com'));
      userEvent.keyboard(',');
      await waitFor(() =>
        expect(
          getByRole('button', { name: 'Edit forem@gmail.com' }),
        ).toBeInTheDocument(),
      );
    });

    it('adds selection when space is pressed', async () => {
      const { getByPlaceholderText, getByRole } = renderMultiInput();
      const input = setup(getByPlaceholderText);

      await waitFor(() => userEvent.type(input, 'forem@gmail.com'));
      userEvent.keyboard(' ');
      await waitFor(() =>
        expect(
          getByRole('button', { name: 'Edit forem@gmail.com' }),
        ).toBeInTheDocument(),
      );
    });
  });

  it('edits a selection', async () => {
    const { getByLabelText, getByPlaceholderText, getByRole } =
      renderMultiInput();
    const input = setup(getByPlaceholderText);
    await waitFor(() => userEvent.type(input, 'forem@gmail.com'));
    userEvent.type(input, '{enter}');
    await waitFor(() =>
      getByRole('button', { name: 'Edit forem@gmail.com' }).click(),
    );

    // Input is focused and pre-filled with value
    const editInput = getByLabelText('Example label');
    await waitFor(() => expect(editInput).toHaveValue('forem@gmail.com'));
    expect(editInput).toHaveFocus();
  });

  it('deletes a selection', async () => {
    const { getByLabelText, getByPlaceholderText, getByRole, queryByRole } =
      renderMultiInput();
    const input = setup(getByPlaceholderText);
    await waitFor(() => userEvent.type(input, 'forem@gmail.com'));
    userEvent.type(input, '{enter}');
    await waitFor(() =>
      getByRole('button', { name: 'Remove forem@gmail.com' }).click(),
    );

    // Selection disappears
    await waitFor(() =>
      expect(
        queryByRole('button', { name: 'Remove forem@gmail.com' }),
      ).not.toBeInTheDocument(),
    );

    // Input is re-focused
    expect(getByLabelText('Example label')).toHaveFocus();
  });

  describe('callbacks', () => {
    it('passes an edit callback to custom selection template', async () => {
      const Selection = ({ name, onEdit }) => (
        <button onClick={onEdit}>Selected: {name}</button>
      );

      const { getByLabelText, getByRole, queryByRole, getByPlaceholderText } =
        render(
          <MultiInput
            labelText="Example label"
            placeholder="Add an email address..."
            inputRegex={/([a-zA-Z0-9@.])/}
            validationRegex={/^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$/}
            SelectionTemplate={Selection}
          />,
        );

      const input = getByPlaceholderText('Add an email address...');
      input.focus();
      await waitFor(() => userEvent.type(input, 'one'));
      userEvent.type(input, '{enter}');
      await waitFor(() =>
        getByRole('button', { name: 'Selected: one' }).click(),
      );

      // Selection disappears
      await waitFor(() =>
        expect(
          queryByRole('button', { name: 'Selected: one' }),
        ).not.toBeInTheDocument(),
      );

      // Input is focused and pre-filled with value
      const editInput = getByLabelText('Example label');
      await waitFor(() => expect(editInput).toHaveValue('one'));
      expect(editInput).toHaveFocus();
    });

    it('passes a deselect callback to custom selection template', async () => {
      const Selection = ({ name, onDeselect }) => (
        <button onClick={onDeselect}>Selected: {name}</button>
      );

      const { getByLabelText, getByRole, queryByRole, getByPlaceholderText } =
        render(
          <MultiInput
            labelText="Example label"
            placeholder="Add an email address..."
            inputRegex={/([a-zA-Z0-9@.])/}
            validationRegex={/^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$/}
            SelectionTemplate={Selection}
          />,
        );

      const input = getByPlaceholderText('Add an email address...');
      input.focus();
      await waitFor(() => userEvent.type(input, 'one'));
      userEvent.type(input, '{enter}');
      await waitFor(() =>
        getByRole('button', { name: 'Selected: one' }).click(),
      );

      // Selection disappears
      await waitFor(() =>
        expect(
          queryByRole('button', { name: 'Selected: one' }),
        ).not.toBeInTheDocument(),
      );

      // Input is re-focused
      expect(getByLabelText('Example label')).toHaveFocus();
    });
  });

  it('edits a previous selection (if exists) on backspace press in empty input', async () => {
    const { getByLabelText, getByRole, queryByRole, getByPlaceholderText } =
      renderMultiInput();

    const input = setup(getByPlaceholderText);
    await waitFor(() => userEvent.type(input, 'example'));
    userEvent.type(input, '{enter}');
    await waitFor(() =>
      expect(getByRole('button', { name: 'Edit example' })).toBeInTheDocument(),
    );

    // Selection should be present as passed as a default value
    const editInput = getByLabelText('Example label');
    editInput.focus();
    await userEvent.type(editInput, '{backspace}');

    await waitFor(() => expect(editInput).toHaveValue('example'));
    expect(queryByRole('button', { name: 'Edit example' })).toBeNull();
  });

  it('Adds a description to invalid entries', async () => {
    const { getByRole, getByPlaceholderText } = renderMultiInput();

    const input = setup(getByPlaceholderText);
    await waitFor(() => userEvent.type(input, 'example'));
    userEvent.type(input, '{enter}');
    await waitFor(() =>
      expect(getByRole('button', { name: 'Edit example' })).toBeInTheDocument(),
    );

    expect(
      getByRole('button', { name: 'Edit example' }),
    ).toHaveAccessibleDescription('Invalid entry');
  });

  it('Does not add a description to valid entries', async () => {
    const { getByRole, getByPlaceholderText } = renderMultiInput();

    const input = setup(getByPlaceholderText);
    await waitFor(() => userEvent.type(input, 'example@email.com'));
    userEvent.type(input, '{enter}');
    await waitFor(() =>
      expect(
        getByRole('button', { name: 'Edit example@email.com' }),
      ).toBeInTheDocument(),
    );

    expect(
      getByRole('button', { name: 'Edit example@email.com' }),
    ).not.toHaveAccessibleDescription('Invalid entry');
  });
});
