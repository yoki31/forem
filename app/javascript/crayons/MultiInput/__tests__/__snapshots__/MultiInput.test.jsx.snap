// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiInput /> renders default UI 1`] = `"<span aria-hidden=\\"true\\" class=\\"absolute pointer-events-none opacity-0 p-2\\"></span><label id=\\"multi-select-label\\">Example label</label><div class=\\"screen-reader-only\\"><p>Selected items:</p><ul aria-live=\\"assertive\\" aria-atomic=\\"false\\" aria-relevant=\\"additions removals\\" class=\\"screen-reader-only list-none\\"></ul></div><div class=\\"c-input--multi relative\\"><div class=\\"c-input--multi__wrapper-border crayons-textfield flex items-center cursor-text pb-9\\"><ul class=\\"list-none flex flex-wrap w-100\\"><li style=\\"order: 1;\\" class=\\"self-center\\"><input autocomplete=\\"off\\" type=\\"text\\" aria-labelledby=\\"multi-select-label\\" placeholder=\\"Add an email address...\\" class=\\"c-input--multi__input\\"></li></ul></div></div>"`;
